### ISSUE TITLE
could you provide a complete code of Diffusion Model with Cross Attention (student).ipynb ？

### ISSUE THREAD

**Issue by mannaandpoem** [Animadversio/DiffusionFromScratch#1]
I learn a lot about stable diffusion from Diffusion Model with Cross Attention (student).ipynb , but I have some troubles on cross Attention when I add the complete code in the space. 
besides，I have a question ：Why is there no gradient term in the equation for forward diffusion, but there is a gradient term in the equation for backward diffusion? 
![image](https://user-images.githubusercontent.com/52203545/227512155-43391a3a-2cfc-4ba2-a041-fb6f1112dc47.png)
![image](https://user-images.githubusercontent.com/52203545/227512169-27c81486-182a-464f-88e1-9a089e1e47b7.png)

**Comment by Animadversio** [comment]
Hi, 
Thanks for noticing my repo! 
I can answer your second question first! Indeed, it's the part the intrigues me the most at first, and it's the key for the diffusion model. 

Using physical intuition, the forward diffusion adds noise, so entropy increases, and information is lost; if backward diffusion just has the 3rd term (which adds the noise) then it will also cause entropy to increase and information will be lost. 
That's why we need the 2nd term to guide the diffusion to climb the "mountain" of log probability, which counter act the diffusing term. 

More mathematically, this term comes from the general formulation of forward and backward SDE, which is for any SDE in the form of 
![image](https://user-images.githubusercontent.com/28660472/228712964-0c84315b-aa75-42d0-a86f-195f480312de.png)
Then there is a reverse process that takes it backward in time 
![image](https://user-images.githubusercontent.com/28660472/228713026-dc9bca5e-ac79-4028-b4b3-5eb20d7c4b9b.png)
i.e. when the reverse process has the initial condition of $p_T(x)$ then running it backward will result in the same marginal density at any time $t$

People seem to attribute this result to this 1982 paper by **Brian D.O. Anderson** https://www.sciencedirect.com/science/article/pii/0304414982900515

A more detailed derivation and verbose note that I like is this one https://ludwigwinkler.github.io/blog/ReverseTimeAnderson/
See also this one https://www.vanillabug.com/posts/sde/

Let me know if you have more questions~

**Comment by Animadversio** [comment]
I just updated the readme page and added the link to the teacher version of the notebook. 
Hope you like it.
https://colab.research.google.com/drive/1_MEFfBdOI06GAuANrs1b8L-BBLn3x-ZJ?usp=sharing

**Comment by mannaandpoem** [comment]
> Hi, Thanks for noticing my repo! I can answer your second question first! Indeed, it's the part the intrigues me the most at first, and it's the key for the diffusion model.
> 
> Using physical intuition, the forward diffusion adds noise, so entropy increases, and information is lost; if backward diffusion just has the 3rd term (which adds the noise) then it will also cause entropy to increase and information will be lost. That's why we need the 2nd term to guide the diffusion to climb the "mountain" of log probability, which counter act the diffusing term.
> 
> More mathematically, this term comes from the general formulation of forward and backward SDE, which is for any SDE in the form of ![图像](https://user-images.githubusercontent.com/28660472/228712964-0c84315b-aa75-42d0-a86f-195f480312de.png) Then there is a reverse process that takes it backward in time ![图像](https://user-images.githubusercontent.com/28660472/228713026-dc9bca5e-ac79-4028-b4b3-5eb20d7c4b9b.png) i.e. when the reverse process has the initial condition of pT(x) then running it backward will result in the same marginal density at any time t
> 
> People seem to attribute this result to this 1982 paper by **Brian D.O. Anderson** https://www.sciencedirect.com/science/article/pii/0304414982900515
> 
> A more detailed derivation and verbose note that I like is this one https://ludwigwinkler.github.io/blog/ReverseTimeAnderson/ See also this one https://www.vanillabug.com/posts/sde/
> 
> Let me know if you have more questions~

Thank you very much. Your explanation is very easy  to understand. This allow me to understand the internal implementation detail of the diffusion model.

**Comment by Guanbin-Huang** [comment]
Hi! Animadversio

Firstly, I'd like to commend you on your work; it's truly impressive! I've been closely examining the sample_X_and_score experiment in the ReverseSDE_Diffusion.ipynb notebook and have a few clarifications to seek.

In your approach to approximate the analytical score with a Neural Network, I noticed that while constructing the data-ground-truth pair, there doesn't seem to be any normalization applied to the ground truth of the score. However, in the code snippet, there appears to be a form of normalization in the fourth line. This has left me a bit puzzled. Specifically, while y_pred is normalized, y_train (which serves as the ground truth) isn't normalized by the timestep. This gives an impression that they might be on different scales.

-------- code ----------
for ep in pbar:
      y_pred = score_model_analy(X_train, T_train) 
      loss = torch.mean(torch.sum((y_pred - y_train)**2 * std_vec[:, None], dim=(1))) # <======= here
-------- code ----------
Moreover, in the forward method, the output is normalized:

-------- code ----------
  def forward(self, x, t):
         t_embed = self.embed(t)
          pred = self.net(torch.cat((x,t_embed),dim=1))
          pred = pred / self.marginal_prob_std_f(t)[:, None,]. # <======= here
    return pred
-------- code ----------

Given this, I'm curious about how arithmetic operations can be applied on y_pred and y_train if they're potentially on different scales.

I've also taken a look at the Colab notebook you provided ([link](https://colab.research.google.com/drive/1e2LXHvvufA3thNvdmsEZdLAI5xkjaLVl#scrollTo=ZyxlQxAUbJZb)) and attached an image for reference:
<img width="723" alt="image" src="https://github.com/Animadversio/DiffusionFromScratch/assets/30463291/af59ea74-59a6-43ea-89f0-53fecc5ec596">
I'd greatly appreciate your insights on this matter. Thank you for your time and patience.

Warm regards,
